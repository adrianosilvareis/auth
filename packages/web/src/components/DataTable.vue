<template>
  <div class="form-floating mx-3 mt-3">
    <input
      id="search"
      class="form-control"
      type="text"
      placeholder="Search"
      v-model="search"
    />
    <label for="search">Search</label>
  </div>
  <table class="table">
    <thead>
      <tr>
        <th scope="col" v-for="(header, index) in headers" :key="index">
          {{ header.title }}
        </th>
      </tr>
    </thead>
    <tbody>
      <tr v-for="(item, index) in filterdItems" :key="index">
        <td v-for="(header, index) in headers" :key="index">
          {{ item[header.key] }}
        </td>
      </tr>
    </tbody>
  </table>
</template>

<script lang="ts">
import { Vue, Options } from "vue-class-component";

interface IHeaders<T> {
  title: string;
  key: keyof T;
}

@Options({
  name: "DataTable",
  props: {
    items: { type: Array, required: true },
    headers: { type: Array, required: true },
  },
})
export default class DataTable<T> extends Vue {
  items!: Array<T>;
  headers!: IHeaders<T>[];

  search = "";

  get filterdItems(): T[] {
    return this.items.filter((item) =>
      this.headers.some((header) =>
        item[header.key]
          .toString()
          .toUpperCase()
          .includes(this.search.toLocaleUpperCase())
      )
    );
  }
}
</script>
