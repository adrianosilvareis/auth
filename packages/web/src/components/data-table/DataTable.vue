<template>
  <div class="row">
    <div class="col">
      <div class="form-floating mx-3 mt-3">
        <InputSearch v-model:search="search" />
      </div>
    </div>
    <div class="col d-flex justify-content-end">
      <div>button</div>
    </div>
  </div>
  <table class="table">
    <thead>
      <tr>
        <th scope="col" v-for="(header, index) in headers" :key="index">
          {{ header.title }}
        </th>
        <th v-if="hasActions">actions</th>
      </tr>
    </thead>
    <tbody>
      <tr v-for="(item, index) in filteredItems" :key="index">
        <td v-for="(header, index) in headers" :key="index">
          <span v-if="editable && index === 0">
            <router-link :to="`${currentRouter}/${item[header.key]}`">
              {{ item[header.key] }}
            </router-link>
          </span>
          <span v-else>
            {{ item[header.key] }}
          </span>
        </td>
        <td v-if="hasActions">
          <slot name="actions" :item="item">
            <BtnRemove v-if="removable" @click="remove(item)" />
          </slot>
        </td>
      </tr>
      <tr v-if="notItems">
        <td class="text-center text-nowrap text-muted">
          <span v-if="search">Nenhum item encontrado para "{{ search }}"!</span>
          <span v-else>Nenhum item encontrado!</span>
        </td>
      </tr>
    </tbody>
  </table>
</template>

<script lang="ts">
import { Vue, Options } from "vue-class-component"
import { ItemModel, Headers, ColumnsConfig } from "./protocols/data-table-utils"
import InputSearch from "@/components/form/InputSearch.vue"
import BtnRemove from "@/components/btn/BtnRemove.vue"

@Options({
  name: "DataTable",
  components: {
    InputSearch,
    BtnRemove
  },
  props: {
    items: { type: Array, required: true },
    columns: { type: Array, required: true },
    editable: { type: Boolean },
    removable: { type: Boolean }
  },
  emits: ["remove-item"]
})
export default class DataTable<T extends ItemModel> extends Vue {
  items!: Array<T>
  columns!: ColumnsConfig<T>
  editable!: boolean
  removable!: boolean

  search = ""
  get headers(): Headers<T>[] {
    return this.columns.map((column): Headers<T> => {
      if (typeof column === "string") {
        return {
          title: column,
          key: column
        }
      }
      if (typeof column === "object") {
        return {
          title: column.title,
          key: column.key
        }
      }
    })
  }

  get notItems(): boolean {
    return !this.filteredItems.length
  }

  get hasActions(): boolean {
    return !!this.$slots.actions || this.removable
  }

  get filteredItems(): T[] {
    if (!this.search) return this.items
    return this.items.filter((item) =>
      this.headers.some((header) =>
        item[header.key]
          .toString()
          .toUpperCase()
          .includes(this.search.toLocaleUpperCase())
      )
    )
  }

  get currentRouter(): string {
    return this.$route.fullPath
  }

  remove(item: T): void {
    this.$emit("remove-item", item)
  }
}
</script>
