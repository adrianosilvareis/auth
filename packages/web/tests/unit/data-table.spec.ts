import { shallowMount } from "@vue/test-utils";
import DataTable from "@/components/data-table/DataTable.vue";

describe("DataTable.vue", () => {
  it.todo('smock tests');
  // it("renders props.headers and items values when passed", () => {
  //   const { wrapper, items, headers } = makeWrapper();
  //   expect(wrapper.text()).toMatch(headers[0].title);
  //   expect(wrapper.text()).toMatch(headers[1].title);
  //   expect(wrapper.text()).toMatch(items[0]["myKey"]);
  //   expect(wrapper.text()).toMatch(items[0]["myKey2"]);
  //   expect(wrapper.element).toMatchSnapshot();
  // });

  // it("renders items filtered when search not null", () => {
  //   const { wrapper, items, headers } = makeWrapper();
  //   wrapper.setData({ search: "fil" });
  //   expect(wrapper.text()).toMatch(headers[0].title);
  //   expect(wrapper.text()).toMatch(headers[1].title);
  //   expect(wrapper.text()).not.toMatch(items[0]["myKey"]);
  //   expect(wrapper.text()).not.toMatch(items[0]["myKey2"]);
  //   expect(wrapper.text()).toMatch(items[1]["myKey"]);
  //   expect(wrapper.text()).toMatch(items[1]["myKey2"]);
  //   expect(wrapper.element).toMatchSnapshot();
  // });
});

function makeWrapper() {
  const headers = makeHeaders();
  const items = makeItems();
  const wrapper = shallowMount(DataTable, {
    props: { headers, items },
  });

  return {
    headers,
    items,
    wrapper,
  };
}

type MyMockTest = {
  myKey: string;
  myKey2: string;
};

function makeItems(): MyMockTest[] {
  return [
    {
      myKey: "any-value",
      myKey2: "any-value-2",
    },
    {
      myKey: "filtered",
      myKey2: "filtered-2",
    },
  ];
}

interface HeaderDataTable {
  title: string;
  key: string;
}
function makeHeaders(): HeaderDataTable[] {
  return [
    {
      title: "Column Label",
      key: "myKey",
    },
    {
      title: "Column Label 2",
      key: "myKey2",
    },
  ];
}
