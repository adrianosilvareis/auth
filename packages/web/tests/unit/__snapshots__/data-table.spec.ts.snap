// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable.vue renders items filtered when search not null 1`] = `
<div
  data-v-app=""
>
  
  <div
    class="form-floating mx-3 mt-3"
  >
    <input
      class="form-control"
      id="search"
      placeholder="Search"
      type="text"
    />
    <label
      for="search"
    >
      Search
    </label>
  </div>
  <table
    class="table"
  >
    <thead>
      <tr>
        
        <th
          scope="col"
        >
          Column Label
        </th>
        <th
          scope="col"
        >
          Column Label 2
        </th>
        
      </tr>
    </thead>
    <tbody>
       [
  {
    "myKey": "any-value",
    "myKey2": "any-value-2"
  },
  {
    "myKey": "filtered",
    "myKey2": "filtered-2"
  }
] 
      
      <tr>
        
        <td>
          any-value
        </td>
        <td>
          any-value-2
        </td>
        
      </tr>
      <tr>
        
        <td>
          filtered
        </td>
        <td>
          filtered-2
        </td>
        
      </tr>
      
    </tbody>
  </table>
  
</div>
`;
